pkg/observatory/observatory_sink.go

```diff
@@ -102,27 +125,21 @@ func (s *ObservatorySink) WriteImmediately(msgs []logger.Message) error {
 	return nil
 }
 
-func (s *ObservatorySink) Flush() error {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	return s.flush()
-}
-
 func (s *ObservatorySink) flush() error {
-	if len(s.buffer) == 0 {
-		return nil
-	}
+	for key, buffer := range s.buffers {
+		buffer.mutex.Lock()
```

Comment:
since buffer is of a pointer type, you may get Nil pointer here, under certain circumstances